src/tests/testClpMap.cc

```diff
@@ -0,0 +1,346 @@
+        m.add("new-key", 0);
+        CPPUNIT_ASSERT_EQUAL(static_cast<size_t>(11), m.entries());
+    }
+    {
+        Map m(1024, 5);
+        addSequenceOfEntriesToMap(m, 1000, 0, 10);
+        CPPUNIT_ASSERT(m.entries() < 1000);
+    }
+}
+
+void
+TestClpMap::testMemoryCounter()
+{
+    CPPUNIT_ASSERT_EQUAL(sizeof(int), static_cast<size_t>(DefaultMemoryUsage(int{})));
+    CPPUNIT_ASSERT_EQUAL(sizeof(int32_t), static_cast<size_t>(DefaultMemoryUsage(int32_t{})));
+    CPPUNIT_ASSERT_EQUAL(sizeof(int64_t), static_cast<size_t>(DefaultMemoryUsage(int64_t{})));
+    CPPUNIT_ASSERT_EQUAL(sizeof(char), static_cast<size_t>(DefaultMemoryUsage(char{})));
+    using Str = char[10];
+    CPPUNIT_ASSERT_EQUAL(sizeof(Str), static_cast<size_t>(DefaultMemoryUsage(Str{})));
+}
+
+void
+TestClpMap::testConstructor()
+{
+    const Map nilA(0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilA.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilA.entries());
+
+    const Map nilB(0, 0);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), nilB.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), nilB.entries());
+
+    const Map emptyC(1);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1), emptyC.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyC.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyC.entries());
+
+    const Map emptyD(1024);
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.memLimit());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(1024), emptyD.freeMem());
+    CPPUNIT_ASSERT_EQUAL(uint64_t(0), emptyD.memoryUsed());
+    CPPUNIT_ASSERT_EQUAL(size_t(0), emptyD.entries());
+}
+
+void
+TestClpMap::testMemoryLimit()
+{
+    const size_t initialCapacity = 1024; // bytes
+    Map m(initialCapacity);
+    fillMapWithEntries(m);
+    const auto entriesAtInitialCapacity = m.entries();
+
+    // check that all entries are removed if we prohibit storage of any entries
+    m.setMemLimit(0);
+    CPPUNIT_ASSERT_EQUAL(size_t(0), m.entries());
+
+    // test whether the map can grow after the all-at-once purging above
+    m.setMemLimit(initialCapacity * 2);
+    fillMapWithEntries(m);
+    CPPUNIT_ASSERT(m.entries() > entriesAtInitialCapacity);
+
+    // test that memory usage and entry count decrease when the map is shrinking
+    while (m.entries()) {
+        // TODO: Check that we can still add a (smaller) entry here.
+
+        const auto memoryUsedBefore = m.memoryUsed();
+        const auto entriesBefore = m.entries();
+
+        const auto newMemoryLimit = memoryUsedBefore/2; // may become zero
+        m.setMemLimit(newMemoryLimit);
+
+        CPPUNIT_ASSERT(m.memoryUsed() <= newMemoryLimit);
+        CPPUNIT_ASSERT(m.entries() < entriesBefore);
+    }
+
```

Comment:
paired with
```suggestion
    CPPUNIT_ASSERT(maxIterations > 0);
```

Reply:
If (and only if) you decide to add those maxIteration checks despite my concerns:

```suggestion
    assert(!m.entries()); // if this fails, the above loop is broken (and not necessarily ClpMap)
 
```

or, if you insist,

```suggestion
    assert(maxIterations > 0); // if this fails, the above loop is broken (and not necessarily ClpMap)

```

or something like that.

Reply:
By "concerns" do you mean the fact that the test - and not the map - might be broken?
My concern is similar; I'd like to ensure that the test reliably terminates, that's less confusing than the alternative.

We can go one step further and use CPPUNIT_ASSERT_MESSAGE(message, condition) to print the diagnostic out in the test output itself

e.g.
```
CPPUNIT_ASSERT_MESSAGE("endless loop detected, aborting", !m.entries() && maxIterations > 0)
```
or something similar

Reply:
> By "concerns" do you mean the fact that the test - and not the map - might be broken? 

No, I did not mean that. I was concerned that "adding more code, adding more implicit dependencies among constants, and making the loop condition more complex" does more harm than good in this specific case. See https://github.com/squid-cache/squid/pull/1262#discussion_r1116440482 for the associated discussion.

Most of those concerns have not changed. However, I was wrong that this test loop is implemented correctly! I fixed my  bug in branch commit a39baed. Hopefully, that fix addresses your concerns as well. The ball is on your side.